src/radical/pilot/agent/lm/spark.py

```diff
@@ -0,0 +1,285 @@
+                    spark_conf_slaves.write('\n')
+
+            spark_conf_slaves.close()
+
+            ## put Master Ip in spark-env.sh file - 
+
+            python_path = os.getenv('PYTHONPATH')
+            python = ru.which('python')
+            logger.info('Python Executable: %s' % python)
+            master_ip = lrms.node_list[0]+hostname
+
+            #Setup default env properties:
+            spark_default_file = open(spark_home + "/conf/spark-defaults.conf",'w')
+            spark_master_string = 'spark://%s:7077' % master_ip
+            spark_default_file.write('spark.master  ' + spark_master_string + '\n')
+            spark_default_file.close()
+            logger.info("Let's print the config")
+            logger.info('Config : {0}'.format(cfg['resource_cfg']))
+
+            spark_env_file = open(spark_home + "/conf/spark-env.sh",'w')
+            #load in the spark enviroment of master and slaves the
+            #configurations of the machine
+            if master_ip!='localhost':
+                for config in cfg['resource_cfg']['pre_bootstrap_1']:
+                    spark_env_file.write(config + '\n')
+
+            spark_env_file.write('export SPARK_MASTER_IP=' + master_ip + "\n")
+            spark_env_file.write('export SCALA_HOME='+ scala_home+ "\n")
+            spark_env_file.write('export JAVA_HOME=' + java_home + "\n")
+            spark_env_file.write('export SPARK_LOG_DIR='+os.getcwd()+'/spark-logs'+'\n')
+            spark_env_file.write('export PYSPARK_PYTHON='+python+'\n')
+
+            spark_env_file.close()
+
+
+            #### Start spark Cluster
+            spark_start = subprocess.check_output(spark_home + '/sbin/start-all.sh')
+            if 'Error' in spark_start:
+                raise RuntimeError("Spark Cluster failed to start: %s" % spark_start)
+            else:
+                logger.info('Start Spark Cluster')
+            launch_command = spark_home +'/bin'
+
+          
+        # The LRMS instance is only available here -- everything which is later
+        # needed by the scheduler or launch method is stored in an 'lm_info'
+        # dict.  That lm_info dict will be attached to the scheduler's lrms_info
+        # dict, and will be passed around as part of the opaque_slots structure,
+        # so it is available on all LM create_command calls.
+        lm_info = {'spark_home'    : spark_home,
+                   'master_ip'     : master_ip,
+                   'lm_detail'     : spark_master_string,
+                   'name'          : lrms.name,
+                   'launch_command': launch_command,
+                   'nodename'      : lrms.node_list[0]}
+
+        return lm_info
+
+
+    # --------------------------------------------------------------------------
+    #
+    @classmethod
+    def lrms_shutdown_hook(cls, name, cfg, lrms, lm_info, logger):
+        if 'name' not in lm_info:
+            raise RuntimeError('rm_ip not in lm_info for %s' \
```

Comment:
The error message complains about `rm_ip`, the check is for `name` -- which is it?